lib/compat.php

```diff
@@ -304,6 +304,247 @@ function gutenberg_warn_classic_about_blocks() {
+		return $html;
+	}
+
+	$defaults = array(
+		'url'   => '',
+		'alt'   => '',
+		'id'    => 0,
+		'align' => '',
+	);
+
+	$block_attributes = wp_parse_args( $block['attrs'], $defaults );
+
+	if ( empty( $block_attributes['url'] ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	if ( ! empty( $block_attributes['id'] ) ) {
+		$attachment_id = (int) $block_attributes['id'];
+		$image_meta    = wp_get_attachment_metadata( $attachment_id );
+	} else {
+		$image_meta    = null;
+		$attachment_id = 0;
+	}
+
+	$image_dimensions = gutenberg_get_image_width_height( $block_attributes, $image_meta );
+	$image_src        = '';
+	$srcset           = '';
+	$sizes            = '';
+
+	if ( empty( $image_dimensions ) ) {
+		// We don't have enough data to construct new img tag. Fall back to the existing HTML.
+		return $html;
+	}
+
+	$image_src = $block_attributes['url'];
+
+	$image_attributes = array(
+		'src'    => $image_src,
+		'alt'    => empty( $block_attributes['alt'] ) ? '' : $block_attributes['alt'],
+		'width'  => $image_dimensions[0],
+		'height' => $image_dimensions[1],
+	);
+
+	if ( $image_meta ) {
+		// TODO: pass `$block_attributes` to the filter.
+		$srcset = wp_calculate_image_srcset( $image_dimensions, $image_src, $image_meta, $attachment_id );
+
+		if ( ! empty( $srcset ) ) {
+			// TODO: pass `$block_attributes` to the filter. This will let themes generate better `sizes` attribute.
+			$sizes = wp_calculate_image_sizes( $image_dimensions, $image_src, $image_meta, $attachment_id );
+		}
+
+		if ( $srcset && $sizes ) {
+			$image_attributes['srcset'] = $srcset;
+			$image_attributes['sizes']  = $sizes;
+		}
+	}
+
+	/**
+	 * Filters the image tag attributes when rendering the core image block.
+	 *
+	 * @since 4.x.x
+	 *
+	 * @param array $image_attributes The (recalculated) image attributes.
+	 *                                Note: expects valid HTML 5.0 attribute names.
+	 * @param array $block_attributes The image block attributes.
+	 * @param string $html The image block HTML coming from the editor. The img tag will be replaced.
+	 */
+	$image_attributes = apply_filters( 'render_block_core_image_tag_attributes', $image_attributes, $block_attributes, $html );
+
+	$attr = '';
+	foreach ( $image_attributes as $name => $value ) {
+		// Sanitize for valid HTML 5.0 attribute names, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes.
+		$name = preg_replace( '/[^a-z0-9-]+/', '', strtolower( $name ) );
+
+		if ( empty( $name ) ) {
+			continue;
+		}
+
+		if ( 'src' === $name ) {
+			$value = esc_url( $value );
+		} elseif ( ( 'width' === $name || 'height' === $name ) && ! empty( $value ) ) {
+			$value = (int) $value;
+		} else {
+			$value = esc_attr( $value );
+		}
+
+		if ( empty( $value ) ) {
+			$attr .= ' ' . $name;
+		} else {
+			$attr .= sprintf( ' %s="%s"', $name, $value );
+		}
+	}
+
+	$image_tag = '<img' . $attr . '/ >';
```

Comment:
Noting that this seems like it could serve as a generally reusable tag builder, not too much unlike [`wp.element.renderToString`](https://github.com/WordPress/gutenberg/blob/f91e66ca2bcd9e1e7b97aa39c01f2888fc751712/packages/element/src/serialize.js#L425)